# Исследование надёжности заёмщиков

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Первый-взгляд" data-toc-modified-id="Первый-взгляд-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Первый взгляд</a></span></li><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обработка пропусков</a></span></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Замена типа данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Лемматизация</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Ответы-на-поставленные-вопросы" data-toc-modified-id="Ответы-на-поставленные-вопросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ответы на поставленные вопросы</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Есть ли зависимость между наличием детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

## Первый взгляд

In [ ]:
import pandas as pd
from pymystem3 import Mystem

In [1]:
df = pd.read_csv('/datasets/data.csv')
df.info()
df.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

**Вывод**

В исходной таблице 12 столбцов, из которых наиболее интересных для исследования три - 'children', 'family_status_id' и 'debt'. Также, в данных присутствуют пропуски, однако только в двух столбцах, хранящих в себе данные о трудовом стаже и размере зарплаты. Причём количество пропусков в этих столбцах совпадает, из чего можно сделать предположение, что нехватает данных в одних и тех же строках. В этих же столбцах данные требуют дополнительной обработки - значения должны быть положительными и целочисленными.
Каждая строка таблицы содержит данные об одном клиенте банка, описывая его семейное положение, наличие детей, данные о работе и образовании. Практически в половине столбцов тип данных object, что может означать высокий риск ошибки в следствие человеческого фактора - необходимо проверить данные на опечатки. Первых пяти строк таблицы достаточно, чтобы заметить некорректный регистр в некоторых ячейках. 

## Подготовка данных

### Обработка пропусков

In [2]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Суммарное количество пропусков в столбцах 'days_employed' и 'total_income' одинаковое - 2174. Теперь можно проверить, в разных они строках или нет.

In [3]:
print(len(df) - len(df.dropna(subset=['days_employed', 'total_income'], inplace=False)))

2174


Разница в количестве строк между изначальной таблицей и таблицей без строк с пропусками - 2174. Это означает, что информации о трудовом стаже и размере зарплаты нет у одних и тех же людей. 

In [5]:
print('Людей, у которых 0 дней трудового стажа:', df[df['days_employed'] == 0]['days_employed'].count())
print('Людей, у которых месячная зарплата равна 0:', df[df['total_income'] == 0]['total_income'].count())

Людей, у которых 0 дней трудового стажа: 0
Людей, у которых месячная зарплата равна 0: 0


В таблице нет людей, у которых в столбцах 'days_employed' или 'total_income' стояло бы значение 0. Скорее всего, в данных оказались пропуски либо в случае отсутствия у человека работы (официально), либо если он отказался предоставить данные. Удалить строки нельзя - будут потеряны важные данные. Поэтому, чтобы избежать проблем с анализом, заменим пропуски на медианное значение.

In [6]:
print('Медианный трудовой стаж:', df['days_employed'].median())
print('Медианная месячная зарплата:', df['total_income'].median())

Медианный трудовой стаж: -1203.369528770489
Медианная месячная зарплата: 145017.93753253992


In [7]:
df['days_employed'] = df['days_employed'].fillna(df['days_employed'].median())
df['total_income'] = df['total_income'].fillna(df['total_income'].median())
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Вывод**

Все пропуски в таблице касались работы клиентов - общий трудовой стаж и ежемесячный доход. Пропуски могли возникнуть по причине отсутствия официальной работы у клиента, или же он отказался предоставить необходимую информацию. Однако, остальные столбцы в этих строках несут в себе полезную информацию, и, чтобы её не потерять, пропущенные данные были заполнены медианными значениями.

### Замена типа данных

In [8]:
df['days_employed'] = df['days_employed'].astype(int)
df['total_income'] = df['total_income'].astype(int)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


In [9]:
print('Количество отрицательных значений трудового стажа:', df[df['days_employed'] < 0]['days_employed'].count(), 'из', len(df))

Количество отрицательных значений трудового стажа: 18080 из 21525


In [10]:
df.sort_values(by='days_employed').head()

children  days_employed  dob_years education  education_id  \
16335         1         -18388         61   среднее             1   
4299          0         -17615         61   среднее             1   
7329          0         -16593         60    высшее             0   
17838         0         -16264         59   среднее             1   
16825         0         -16119         64   среднее             1   

         family_status  family_status_id gender income_type  debt  \
16335  женат / замужем                 0      F   сотрудник     0   
4299   женат / замужем                 0      F   компаньон     0   
7329   женат / замужем                 0      F   сотрудник     0   
17838  женат / замужем                 0      F   сотрудник     0   
16825  женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
16335        186178      операции с недвижимостью  
4299         122560                 покупка жилья  
7329         124697  заняться высшим образованием  
17838         51238         на покупку автомобиля  
16825         91527    покупка жилой недвижимости

In [11]:
df.sort_values(by='days_employed').tail()

children  days_employed  dob_years education  education_id  \
7794          0         401663         61   среднее             1   
2156          0         401674         60   среднее             1   
7664          1         401675         61   среднее             1   
10006         0         401715         69    высшее             0   
6954          0         401755         56   среднее             1   

               family_status  family_status_id gender income_type  debt  \
7794        гражданский брак                 1      F   пенсионер     0   
2156         женат / замужем                 0      M   пенсионер     0   
7664         женат / замужем                 0      F   пенсионер     0   
10006  Не женат / не замужем                 4      F   пенсионер     0   
6954          вдовец / вдова                 2      F   пенсионер     0   

       total_income                purpose  
7794          48286                свадьба  
2156         325395             автомобили  
7664         126214      операции с жильем  
10006         57390  получение образования  
6954         176278           ремонт жилью

**Вывод**

Все столбцы с числовыми значениями приведены к целочисленному типу методом astype(), поскольку дробной частью числа в данном случае можно пренебречь. Также проведено сравнение максимальных и минимальных значений столбца 'days_employed'. Можно сделать вывод, что наличие отрицательных значений - не случайная ошибка, и просто отбросить знак нельзя, поскольку разница крайних значений по модулю слишком велика. Напрашивается желание уточнить природу артефакта у того, кто эти данные предоставил. Возможно, такие значения получились из-за различий в системах подсчёта стажа. Если разные работодатели указывают значения в разном формате, то при подведении итога могли возникнуть отрицательные знчения. 

### Обработка дубликатов

In [12]:
df.duplicated().sum()

54

In [13]:
df['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

In [14]:
df['education'] = df['education'].str.lower()
df['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

In [15]:
df.duplicated().sum()

71

In [16]:
df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [17]:
df['family_status'] = df['family_status'].str.lower()
df['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64

In [18]:
df['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
жилье                                     647
покупка жилья                             647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

In [19]:
df = df.drop_duplicates().reset_index(drop=True)
df.duplicated().sum()

0

**Вывод**

Изначальное количество дубликатов в таблице - 54. Однако проверка показала, что в столбце 'education' много повторов с изменением регистра. После приведения всех значений столбца к одному виду количество дубликатов в таблице увеличилось до 71. В столбце 'family_status' тоже были значения с некорректным регистром, однако это не влияет на поиск дубликатов. Последний столбец, 'purpose', содержит в себе много повторяющихся по смыслу значений, которые необходимо разделить на группы. Возможная причина возникновения дубликатов: клиент мог подать заявление повторно, не дождавшись ответа или получив отказ. Например, заявка может составляться не напрямую в банке, а у партнёра, например в каком-то магазине или дилерском центре. В таком случае клиент мог подать несколько заявок в нескольких магазинах. Также, некоторое количество дубликатов могло возникнуть после того, как был изменён тип данных. Мы отбросили дробную часть, в которой могли быть различия.

### Лемматизация

In [20]:
df_clean = df
m = Mystem()
def f_lem(row):
    lem = m.lemmatize(row['purpose'])
    return lem

df_clean['purpose_lem'] = df.apply(f_lem, axis=1)
df_clean['purpose_lem'].value_counts()

[автомобиль, \n]                                          972
[свадьба, \n]                                             791
[на,  , проведение,  , свадьба, \n]                       768
[сыграть,  , свадьба, \n]                                 765
[операция,  , с,  , недвижимость, \n]                     675
[покупка,  , коммерческий,  , недвижимость, \n]           661
[операция,  , с,  , жилье, \n]                            652
[покупка,  , жилье,  , для,  , сдача, \n]                 651
[операция,  , с,  , коммерческий,  , недвижимость, \n]    650
[покупка,  , жилье, \n]                                   646
[жилье, \n]                                               646
[покупка,  , жилье,  , для,  , семья, \n]                 638
[строительство,  , собственный,  , недвижимость, \n]      635
[недвижимость, \n]                                        633
[операция,  , со,  , свой,  , недвижимость, \n]           627
[строительство,  , жилой,  , недвижимость, \n]            624
[покупка

**Вывод**

Чтобы не изменять исходную таблицу, была создана её копия. Таким образом, любые дальнейшие действия не повредят изначальные данные. Результаты лемматизации сформированы в списки, которые составляют новый столбец таблицы: 'purpose_lem'. 

### Категоризация данных

In [28]:
def f_category(row):
    val = row['purpose_lem']
    
    for i in range(len(val)):
        
        if (val[i] == 'коммерческий'):
            return 'коммерческая недвижимость'
        
        if (val[i] == 'жилье') | (val[i] == 'недвижимость') :
            return 'жилая недвижимость'
        
        if val[i] == 'образование':
            return 'образование'
        
        if val[i] == 'автомобиль':
            return 'автомобиль'
        
        if val[i] == 'свадьба':
            return 'свадьба'
        
    return 'без категории'
        
df_clean['purpose_category'] = df.apply(f_category, axis=1)
df_clean = df_clean.drop(['purpose', 'purpose_lem'], axis=1)
df_clean['purpose_category'].value_counts()

жилая недвижимость           9500
автомобиль                   4306
образование                  4013
свадьба                      2324
коммерческая недвижимость    1311
Name: purpose_category, dtype: int64

После анализа лемм было принято решение разделить все цели кредитов на 5 категорий: жилая недвижимость, коммерческая недвижимость, образование, автомобиль и свадьба. Недвижимость разделена на коммерческую и жилую, потому что есть предположение, что эти различия могут повлиять на возврат кредита. Возможно, если клиент берёт кредит на коммерческую недвижимость, то он более ответственен и серьёзнее настроен, и поэтому с бóльшей вероятностью вернёт долг вовремя. Также были удалены лишние столбцы, 'purpose' и 'purpose_lem'.

In [29]:
df_clean.duplicated().sum()

298

После лемматизации выяснилось, что в таблице есть ещё некоторое количество дубликатов. Возможные причиы те же - ошибка при указании цели кредита и подача завявки в разных местах.

In [30]:
df_clean = df_clean.drop_duplicates().reset_index(drop=True)
df_clean.duplicated().sum()

0

In [31]:
df_clean.head(10)

children  days_employed  dob_years education  education_id  \
0         1          -8437         42    высшее             0   
1         1          -4024         36   среднее             1   
2         0          -5623         33   среднее             1   
3         3          -4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0           -926         27    высшее             0   
6         0          -2879         43    высшее             0   
7         0           -152         50   среднее             1   
8         2          -6929         35    высшее             0   
9         0          -2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

     purpose_category  
0  жилая недвижимость  
1          автомобиль  
2  жилая недвижимость  
3         образование  
4             свадьба  
5  жилая недвижимость  
6  жилая недвижимость  
7         образование  
8             свадьба  
9  жилая недвижимость

In [32]:
df_clean = df_clean.sort_values(by='total_income', ascending=False).reset_index(drop=True)

for i in range(len(df_clean)):
    val = (len(df_clean)) / 5
    if i <= val:
        df_clean.loc[i, 'total_income_category'] = 'высокий'
        
    if val < i < (2 * val):
        df_clean.loc[i, 'total_income_category'] = 'выше среднего'
        
    if (2 * val) <= i <= (3 * val):
        df_clean.loc[i, 'total_income_category'] = 'средний'
        
    if (3 * val) <= i <= (4 * val):
        df_clean.loc[i, 'total_income_category'] = 'ниже среднего'
        
    if (4 * val) <= i:
        df_clean.loc[i, 'total_income_category'] = 'низкий'
        

In [33]:
df_clean['total_income_category'].value_counts()

высокий          4232
низкий           4231
средний          4231
ниже среднего    4231
выше среднего    4231
Name: total_income_category, dtype: int64

Помимо категоризации значений столбца с целью кредита, нужно что-то делать со значениями ежемесячного дохода. Слишком много разных чисел, хотя по сути они отличаются незначительно. Было принято решение разделить все значения на пять категорий дохода: высокий, выше среднего, средний, ниже среднего и низкий. Для этого таблица отсортирована по столбцу 'total_inome' по убыванию, а затем разбита на 5 равных частей. Каждой части присваивается нужное значение. Результат можно проверить, выведя пограничные строки:

In [36]:
df_clean.loc[4231:4232]

children  days_employed  dob_years education  education_id  \
4231         0          -7076         52   среднее             1   
4232         0          -3472         57   среднее             1   

        family_status  family_status_id gender income_type  debt  \
4231  женат / замужем                 0      M   компаньон     0   
4232  женат / замужем                 0      M   сотрудник     0   

      total_income    purpose_category total_income_category  
4231        215608  жилая недвижимость               высокий  
4232        215599          автомобиль         выше среднего

In [37]:
df_clean.loc[8462:8463]

children  days_employed  dob_years education  education_id  \
8462         0           -211         30    высшее             0   
8463         2           -973         45   среднее             1   

              family_status  family_status_id gender income_type  debt  \
8462  не женат / не замужем                 4      F   сотрудник     0   
8463       гражданский брак                 1      M   сотрудник     0   

      total_income    purpose_category total_income_category  
8462        157648          автомобиль         выше среднего  
8463        157646  жилая недвижимость               средний

In [38]:
df_clean.loc[12693:12694]

children  days_employed  dob_years education  education_id  \
12693         0          -1601         38   среднее             1   
12694         0         368375         66   среднее             1   

               family_status  family_status_id gender income_type  debt  \
12693        женат / замужем                 0      M   сотрудник     0   
12694  не женат / не замужем                 4      F   пенсионер     0   

       total_income    purpose_category total_income_category  
12693        134335  жилая недвижимость               средний  
12694        134334          автомобиль         ниже среднего

In [39]:
df_clean.loc[16924:16925]

children  days_employed  dob_years education  education_id  \
16924         0          -2644         38    высшее             0   
16925         1          -3428         28   среднее             1   

          family_status  family_status_id gender income_type  debt  \
16924   женат / замужем                 0      F   компаньон     0   
16925  гражданский брак                 1      F   сотрудник     0   

       total_income purpose_category total_income_category  
16924         98028       автомобиль         ниже среднего  
16925         98018          свадьба                низкий

**Вывод**

Лемматизация дала возможность разделить значения столбца 'purpose' на группы по смыслу. По итогу категоризации были выявлены дополнительные дубликаты, незамеченные ранее. Также на группы был разбит столбец 'total_income' - будет куда проще искать взаимосвязи с уровнем дохода, если этих уровней всего 5. 

## Ответы на поставленные вопросы

### Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [40]:
def pivot(data, index, columns, values):
    data_pivot = data.pivot_table(index=[index], columns=[columns], values=[values], aggfunc='count')
    data_pivot.set_axis(['clients_without_debt', 'clients_with_debt'], axis='columns', inplace=True)
    data_pivot['conversion'] = (data_pivot['clients_without_debt']) / (data_pivot.apply('sum', axis=1)) * 100
    data_pivot['conversion'] = data_pivot['conversion'].astype(int)
    return data_pivot.sort_values(by='conversion', ascending=False)

pivot_children = pivot(df_clean, 'children', 'debt', 'education')
pivot_children 

clients_without_debt  clients_with_debt  conversion
children                                                     
 5                         9.0                NaN         100
-1                        46.0                1.0          97
 0                     12792.0             1061.0          92
 3                       302.0               27.0          91
 1                      4318.0              444.0          90
 2                      1846.0              194.0          90
 4                        36.0                4.0          90
 20                       68.0                8.0          89

In [41]:
pivot_children = pivot_children.drop([-1, 20], axis=0)
pivot_children['clients_with_debt'] = pivot_children['clients_with_debt'].fillna(0)
pivot_children

clients_without_debt  clients_with_debt  conversion
children                                                     
5                          9.0                0.0         100
0                      12792.0             1061.0          92
3                        302.0               27.0          91
1                       4318.0              444.0          90
2                       1846.0              194.0          90
4                         36.0                4.0          90

In [42]:
pivot_children_percent = pivot_children.sort_values(by='clients_with_debt', ascending=False)
total_clients = pivot_children['clients_without_debt'].sum() + pivot_children['clients_with_debt'].sum() 
pivot_children_percent['total_clients_count'] = ((pivot_children['clients_without_debt'] + pivot_children['clients_with_debt']) * 100) / total_clients
pivot_children_percent = pivot_children_percent.drop(['clients_without_debt', 'clients_with_debt', 'conversion'], axis=1)
pivot_children_percent

total_clients_count
children                     
0                   65.863167
1                   22.640612
2                    9.699044
3                    1.564209
4                    0.190177
5                    0.042790

**Вывод**

Разница в количестве вовремя возвращённых кредитов между клиентами с детьми и без несущественна - клиенты без детей всего на 2% более ответственны. Информацию про -1 или 20 детей можно считать ошибочной и игнорировать, поскольку таких клиентов очень мало. На показатели клиентов с 5 детьми также можно не ориентироваться, поскольку таких клиентов практически нет. Разница между клиентами с детьми и без несущественна, в среднем 90% людей возвращает долг вовремя. Однако, здесь есть другая закономерность - в количестве клиентов с детьми. Чем больше детей в семье, тем с меньшей вероятностью клиент возьмёт кредит. 65% всех клинентов не имеют детей, 1 ребёнок всего у 22%, 2 - у 9%, и так далее. Даже если принимать во внимание, что семьи с большим количеством детей встречаются сильно реже, чем семьи с одим ребёнком, и сравнивать только клиентов без детей и клиентов с 1 ребёнком, то зависимость сохраняется - клиентов без детей в 3 раза больше. Возможно, далеко не все клиенты могут выдержать кредитную нагрузку в совокупности с содержанием ребёнка. Это можно проверить отдельно, используя данные о клиентах банка, которые вообще не взяли кредит.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [43]:
pivot_family_status = pivot(df_clean, 'family_status', 'debt', 'education')
pivot_family_status 

clients_without_debt  clients_with_debt  conversion
family_status                                                             
вдовец / вдова                          881                 63          93
в разводе                              1108                 85          92
женат / замужем                       11177                929          92
гражданский брак                       3737                388          90
не женат / не замужем                  2514                274          90

In [44]:
 def f_family_cat (row):
        if row['family_status'] != 'женат / замужем':
            return 'не женат / не замужем'
        return 'женат / замужем'
    
df_clean['family_status'] = df_clean.apply(f_family_cat, axis=1)
pivot_family_status_low = pivot(df_clean, 'family_status', 'debt', 'education')
pivot_family_status_low

clients_without_debt  clients_with_debt  conversion
family_status                                                             
женат / замужем                       11177                929          92
не женат / не замужем                  8240                810          91

**Вывод**

Ситуация аналогична предыдущей - разница крайне несущественна, 2-3%. Даже если предположить, что существует только две категории, женат/замужем и не женат/не замужем, то разница между группами всего 1%. Более интересно другое - как мы уже знаем, подавляющее число клиентов с кредитами не имеют детей. Но также мы видим, что подавляющее большинство состоит в браке. Получается, что выплачивать кредит людям проще вдвоём, и целевая аудитория банка - семьи, которые ещё не завели ребёнка.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [45]:
pivot_total_income = pivot(df_clean, 'total_income_category', 'debt', 'education')
pivot_total_income

clients_without_debt  clients_with_debt  conversion
total_income_category                                                     
высокий                                3935                297          92
выше среднего                          3868                363          91
ниже среднего                          3885                346          91
низкий                                 3890                341          91
средний                                3839                392          90

**Вывод**

Как и раньше, разница между клиентами с высоким уровнем дохода и низким минимальна, всё те же 1-2%, закономерности нет. Ориентироваться по этим данным на количество клиентов с тем или иным уровнем дохода нельзя, поскольку мы сами разделили их всех на 5 равных категорий.

### Как разные цели кредита влияют на его возврат в срок?

In [46]:
pivot_purpose_category = pivot(df_clean, 'purpose_category', 'debt', 'education')
pivot_purpose_category

clients_without_debt  clients_with_debt  conversion
purpose_category                                                              
жилая недвижимость                         8621                682          92
коммерческая недвижимость                  1212                 99          92
свадьба                                    2120                186          91
автомобиль                                 3870                402          90
образование                                3594                370          90

**Вывод**

Как и раньше, различия практически отсутствуют, клиенты возвращают кредит вне зависимости от цели, на которую он был взят. Гипотеза о том, что люди, взявшие кредит на коммерческую недвижимость более ответственны и вернут долг раньше не подтвердилась, конверсия этой категориии такая же, как и у жилой недвижимости. Самая малочисленная цель - свадьба, но при этом её конверсия не самая худшая. 

## Общий вывод

Ни одна первоначальная гипотеза не подтвердилась. Количество должников не зависит ни от количества детей, ни от семейного положения, ни от дохода, ни от цели кредита. Их всегда около 10%, в каждой категории. Использовать эти данные для постройки модели кредитного скоринга не получится, они не несут в себе чётких закономерностей. Возможно, для этого будут более полезны данные о доходах и расходах клиента, о кредитной истории, о попытках взять кредит в других банках. Однако, анализ полученных данных показал, что количество клиентов без детей сильно превышает количество клиентов с детьми. И это при том, что подавляющее большинство клиентов состоит в браке. Соответственно, когда в этом браке родится ребёнок, а людям потребуется кредит, то с большой вероятностью они выберут другой банк. Возможно, стоит пересмотреть условия и предложить более выгодные программы, подходящие для этой категории людей. 